experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts

```diff
@@ -100,44 +72,30 @@ export function getProxyForField<TSchema extends TreeFieldSchema>(
 	}
 }
 
-export function getProxyForNode<TSchema extends TreeNodeSchema>(
+export function getOrCreateNodeProxy<TSchema extends TreeNodeSchema>(
 	treeNode: TreeNode,
 ): ProxyNode<TSchema> {
-	const schema = treeNode.schema;
-
-	/**
-	 * Gets a cached proxy for `treeNode` if one was already created and cached.
-	 * Otherwise, creates a new proxy, caches it, and returns it.
-	 */
-	function getOrCreateProxy(createProxy: () => SharedTreeNode): ProxyNode<TSchema> {
-		const cachedProxy = getCachedProxy(treeNode);
-		if (cachedProxy !== undefined) {
-			return cachedProxy as ProxyNode<TSchema>;
-		}
-
-		const proxy = createProxy();
-		cacheProxy(treeNode, proxy);
-		return proxy as ProxyNode<TSchema>;
+	const cachedProxy = tryGetTreeNodeTarget(treeNode);
+	if (cachedProxy !== undefined) {
+		return cachedProxy as ProxyNode<TSchema>;
 	}
 
-	if (schemaIsMap(schema)) {
-		return getOrCreateProxy(() => createMapProxy(treeNode));
-	}
+	const schema = treeNode.schema;
 	if (schemaIsLeaf(schema)) {
 		return treeNode.value as ProxyNode<TSchema>;
 	}
-	const isFieldNode = schemaIsFieldNode(schema);
-	if (isFieldNode || schemaIsObjectNode(schema)) {
-		return getOrCreateProxy(() =>
-			isFieldNode ? createListProxy(treeNode) : createObjectProxy(treeNode, schema),
-		);
+	if (schemaIsMap(schema) && treeNode.is(schema)) {
```

Comment:
Why is the `treeNode.is(schema)` check necessary in these? We're getting `schema` via `treeNode.schema`, so shouldn't that check _always_ succeed?

Reply:
Yea, now that you point that out, it's kind of silly. It narrows the type of `treeNode` to be `ObjectNode`, or `FieldNode`, or `MapNode`, so that the types match up when we call `setTreeNode`. I guess I can just cast the second one though, since as you pointed out, it's redundant.